### PR TITLE
load-fragment: resolve specifiers in BindPaths/BindReadOnlyPaths

### PR DESCRIPTION
**Pull request by dmchurch**
Allow specifiers in bind source/destination directives; this is useful e.g. for templated, chrooted services that should have different directories pulled into the chroot.

### CODE REVIEW

File: src/core/load-fragment.c [hunk 1]
```diff
@@ -3939,7 +3941,14 @@ int config_parse_bind_paths(
                         return 0;
                 }
 
-                s = source;
+                r = unit_full_printf(u, source, &sresolved);
+                if (r < 0) {
+                        log_syntax(unit, LOG_ERR, filename, line, -r,
```
Comment (poettering) [inline src/core/load-fragment.c]: no need to negate "r" here, log_syntax() implicitly calls abs() anyway

File: src/core/load-fragment.c [hunk 2]
```diff
@@ -3970,16 +3979,23 @@ int config_parse_bind_paths(
                                 return 0;
                         }
 
-                        if (!utf8_is_valid(destination)) {
-                                log_syntax_invalid_utf8(unit, LOG_ERR, filename, line, destination);
+                        r = unit_full_printf(u, destination, &dresolved);
+                        if (r < 0) {
+                                log_syntax(unit, LOG_ERR, filename, line, -r,
```
Comment (poettering) [inline src/core/load-fragment.c]: same here

**Comment by dmchurch** [comment]
Oops! I was modeling off an old commit that had since had its syntax changed, and I didn't catch it. Fixed!

File: src/core/load-fragment.c [hunk 3]
```diff
@@ -3939,7 +3941,14 @@ int config_parse_bind_paths(
                         return 0;
                 }
 
-                s = source;
+                r = unit_full_printf(u, source, &sresolved);
+                if (r < 0) {
+                        log_syntax(unit, LOG_ERR, filename, line, r,
+                                   "Failed to resolved specifiers in \"%s\", ignoring: %m", source);
+                        return 0;
```
Comment (evverx) [inline src/core/load-fragment.c:3948]: I wonder why we return `0` here if `unit_full_printf` returns `-ENOMEM`. I think we should call `return log_oom()`.

Comment (dmchurch) [reply src/core/load-fragment.c:3948]: I couldn't say for sure (again, I'm just working off pattern-matching here, I don't know the codebase), but I expect we're already logging the error message through the `%m` in the `log_syntax`, and the other syntax errors return `0` - not to mention that `ENOMEM` is presumably not the only error that `unit_full_printf` can return.

Are you suggesting I should explicitly test for `-ENOMEM` in the return from `unit_full_printf` and handle it separately, like for the calls to `extract_first_word`?

Comment (dmchurch) [reply src/core/load-fragment.c:3948]: I'll note that the existing calls to `unit_full_printf` in this file don't handle `-ENOMEM` specially, they simply lump it in with the other errors into a single `log_syntax` call.

**Comment by evverx** [comment]
I think something should be added to `NEWS`, because we are going to break `BindPaths` containing `%`.

**Comment by dmchurch** [comment]
That's a good point, I never considered that people might have paths with `%` symbols in them. Maybe that should go into NEWS, whenever that gets compiled for r234? How do we make sure an entry eventually makes it in there?

File: src/core/load-fragment.c (hunk 3 above)
Comment (evverx) [reply src/core/load-fragment.c:3948]: > I'll note that the existing calls to unit_full_printf in this file don't handle -ENOMEM specially

You are right. This is the result of https://github.com/systemd/systemd/commit/19f6d710772305610b928bc2678b9d77fe11e770#diff-cc81d855afbe30015a0e769bbf67567fL102.

I think we can fix it later.

**Comment by dmchurch** [comment]
Looks like one of the auto tests failed because of a timeout issue, is there some way to get it to rerun? Also, is there something I need to do to get another look at the review?

**Comment by martinpitt** [comment]
I retried the amd64 test. I also unset the "needs work" label, as it appears you addressed @poettering's review comments.

File: src/core/load-fragment.c (hunk 3 above)
Comment (poettering) [reply src/core/load-fragment.c:3948]: so, strictly speaking, @evverx is right: we should catch ENOMEM here explicitly, and then call log_oom() for it, and return an error, as ENOMEM is not a user issue, but a non-recoverable error. That said, we have been sloppy with this in general, so I figure it's OK if this patch doesn't do this either

**Comment by poettering** [comment]
> Maybe that should go into NEWS, whenever that gets compiled for r234? How do we make sure an entry eventually makes it in there?

The NEWS file may be updated as part of any PR, but we'll also update it right before each release adding everything we missed so far.

I'll merge this PR now, as it looks good the way it is, but I'd be very thankful for a follow-up PR that adds a NEWS entry for this too, in preperation for an eventual v234 release!

**Pull request merged**